src/compiler/crystal/semantic/bindings.cr

```diff
@@ -1,7 +1,29 @@
 module Crystal
+  class ASTNode
+    # A node whose type can be frozen.
+    # When frozen, trying to change the node's type to something
+    # that doesn't match `freeze_type` will produce a compilation error.
+    # This is used for checking a method's return type, a block's
+    # expected type, instance and class vars expected types,
+    # and local vars with a type declaration.
+    module Freezable
+      property freeze_type : Type?
+    end
+
+    # By default nodes don't have their type frozen.
+    def freeze_type
+      nil
+    end
+  end
+
+  {% for name in %w(Block MetaVar MetaTypeVar Def) %}
+    class {{name.id}}
+      include Freezable
+    end
+  {% end %}
```

Comment:
It's nice to have the binding-related things all defined here. But on the other hand, reopening a couple of classes just to include a module and define a property is also adding an extra code location you need to look at for understanding the properties of a AST type.
In `semantics/ast.cr` we're already reopening all those ASTNode classes to add semantics functionality.
Perhaps it could be an option to just drop in `property freeze_type : Type?` (or `include Freezable`) there in the respective places?

Reply:
Done!